cli/disputer.go

```diff
@@ -0,0 +1,397 @@
+				}
+				for _, val := range notif {
+					switch val.Type {
+					case store.HCApply:
+
+						if val.Val.Height() < miner3.WPoStChallengeWindow {
+							// it's too early in the chain to challenge any posts
+							continue
+						}
+
+						// TODO: If the newly applied tipset is built on top of n null tipset(s), we should also check the tipsets WPoStChallengeWindow+[1...n]  epochs ago
+
+						// Since we're waiting WPoStChallengeWindow epochs already, we don't have to worry about confidence of the message -- it's already pretty deep onchain
+						ts, err := api.ChainGetTipSetByHeight(ctx, val.Val.Height()-miner3.WPoStChallengeWindow, val.Val.Key())
+						if err != nil {
+							return xerrors.Errorf("failed to get tipset 60 epochs ago: %w", err), false
+						}
+
+						// TODO: Build a cache of tipsets we've already inspected and continue here if appropriate
+
+						pmsgs, err := api.ChainGetParentMessages(ctx, ts.Cids()[0])
+						if err != nil {
+							return xerrors.Errorf("failed to get parent messages: %w", err), false
+						}
+
+						wpmsgs, err := getSubmitWindowedPoSts(ctx, api, pmsgs, ts.Key())
+						if err != nil {
+							return xerrors.Errorf("failed to get post messages: %w", err), false
+						}
+
+						dpmsgs, err := makeDisputeWindowedPosts(ctx, api, wpmsgs, fromAddr)
+						if err != nil {
+							return xerrors.Errorf("failed to check for disputes: %w", err), false
+						}
+
+						for _, dpmsg := range dpmsgs {
+							fmt.Println("disputing a post at height ", ts.Height())
+							_, err := api.MpoolPushMessage(ctx, dpmsg, mss)
+							if err != nil {
+								return xerrors.Errorf("failed to dispute post message: %w", err), false
+							}
+
+							// TODO: Track / report on message landing on chain?
+						}
+					case store.HCRevert:
+						// do nothing
+					default:
+						return xerrors.Errorf("unexpected head change type %s", val.Type), true
+					}
+				}
+			case <-ticker.C:
+				fmt.Print("Running health check: ")
+
+				cctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
+				if _, err := api.ID(cctx); err != nil {
+					cancel()
+					return xerrors.Errorf("health check failed"), true
+				}
+
+				cancel()
+
+				fmt.Println("Node online")
+			case <-ctx.Done():
+				return nil, true
+			}
+
+			return nil, false
+		}
+
+		for {
+			err, shutdown := disputeLoop()
+			if err != nil && shutdown {
+				fmt.Println("disputer errored, shutting down: ", err)
+				break
+			}
+
+			if err != nil && !shutdown {
+				fmt.Println("disputer errored, continuing to run: ", err)
+			}
+
+			if shutdown {
+				fmt.Println("disputer shutting down")
+				break
+			}
+		}
+
+		return nil
+	},
+}
+
+// for a list of messages and the tsk in which they were executed, returns the _successful_ SubmitWindowedPoSt msgs among them
+func getSubmitWindowedPoSts(ctx context.Context, api lapi.FullNode, msgs []lapi.Message, tsk types.TipSetKey) ([]lapi.Message, error) {
```

Comment:
This won't work due to internal messages. A "good" attacker will prove from, e.g., a multisig to avoid this checker.

I believe the _only_ real way to do this is to:

1. Watch for new miners at every epoch, and track them. Maybe watch for entries in the power table?
2. For all tracked miners, track their proving period.
3. At the end of every epoch, check every miner that is "up" for proving. Unfortunately, this will check 1/60 of the miners every epoch, even if the miner has no live sectors. However, that check _should_ be pretty fast.

Reply:
@Stebalien Yeah, you're right...Reimplemented to use Design 1